langserver/workspace_refs.go

```diff
@@ -0,0 +1,155 @@
+package langserver
+
+import (
+	"context"
+	"fmt"
+	"go/build"
+	"go/token"
+	"log"
+	"sort"
+	"strings"
+	"sync"
+
+	"golang.org/x/tools/go/buildutil"
+
+	opentracing "github.com/opentracing/opentracing-go"
+	"github.com/opentracing/opentracing-go/ext"
+	"github.com/sourcegraph/jsonrpc2"
+	"github.com/sourcegraph/sourcegraph-go/langserver/internal/refs"
+	"github.com/sourcegraph/sourcegraph-go/pkg/lspext"
+)
+
+func (h *LangHandler) handleWorkspaceReference(ctx context.Context, conn JSONRPC2Conn, req *jsonrpc2.Request, params lspext.WorkspaceReferenceParams) ([]lspext.ReferenceInformation, error) {
+	rootPath := h.FilePath(h.init.RootPath)
+	bctx := h.OverlayBuildContext(ctx, h.defaultBuildContext(), !h.init.NoOSFileSystemAccess)
+
+	var pkgPat string
+	if h.init.RootImportPath == "" {
+		// Go stdlib (empty root import path)
+		pkgPat = "..."
+	} else {
+		// All other Go packages.
+		pkgPat = h.init.RootImportPath + "/..."
+	}
+
+	results := refResultSorter{results: make([]lspext.ReferenceInformation, 0)}
+	var wg sync.WaitGroup
+	pkgs := buildutil.ExpandPatterns(bctx, []string{pkgPat})
+	for pkg := range pkgs {
+		wg.Add(1)
+		go func(pkg string) {
+			defer wg.Done()
+			err := h.externalRefsFromPkg(ctx, bctx, conn, pkg, rootPath, &results)
+			if err != nil {
+				log.Printf("externalRefsFromPkg: %v: %v\n", pkg, err)
+			}
+		}(pkg)
+	}
+	wg.Wait()
+
+	sort.Sort(&results) // sort to provide consistent results
+	if len(results.results) > params.Limit && params.Limit > 0 {
+		results.results = results.results[:params.Limit]
```

Comment:
Yes, that is true. The point is not to actually use `Limit` but just to provide it such that other implementations get in the mindset of supporting it, in case we decide down the road that limiting or pagination is needed, it'll be easier if people designed from the beginning to support it. I will add a TODO mentioning this.